jekyll/_cci2/caching.md

```diff
@@ -1,74 +1,172 @@
+      # Find a cache corresponding to this specific package.json checksum
+      # when this file is changed, this key will fail
+      - v1-projectname-npm-deps-{{ .Branch }}-{{ checksum "package.json" }}
+      # Find a cache corresponding to any build in this branch, regardless of package.json
+      # checksum.  The most recent one will be used.
+      - v1-projectname-npm-deps-{{ .Branch }}
+      # Find the most recent cache used from any branch
+      - v1-projectname-npm-deps-
 ```
 {% endraw %}
 
-Note that the keys become less specific as we move through the list.
+Note that the keys become less specific as we move through the list. As caches become less specific there is greater likelihood that the dependencies they contain are different from those that the current job requires. When your dependency tool runs (e.g. `npm install`) it will discover out-of-date dependencies and install the ones the current job specifies. This is 'partial cache' restore.
+
+### Get clean caches when language or dependency management tool versions change
+
+<div class="alert alert-info" role="alert">
+<b>Tip:</b> Caches are immutable so it's useful to start all your cache keys with a version prefix, e.g. <code class="highlighter-rouge">v1-...</code>. That way you will be able to regenerate all your caches by incrementing the version in this prefix.
+</div>
+
+Using a strategy similar to what is shown above, you can 'clear' your cache when you change things in your project that need a new cache. For example:
+
+* dependency manager version change (e.g. you change npm from 4 -> 5)
+* language version change (e.g. you change ruby 2.3 -> 2.4)
+* dependencies removed from your project
+
+### `save_cache` template examples:**
+
+Now you understand how cache is restored we can learn how to save caches. Here are some examples of strategies we've found to be useful:
+
+ * {% raw %}`myapp-{{ checksum "package.json" }}`{% endraw %} - cache will be regenerated every time something is changed in `package.json` file, different branches of this project will generate the same cache key.
+ * {% raw %}`myapp-{{ .Branch }}-{{ checksum "package.json" }}`{% endraw %} - same as the previous one, but each branch will generate separate cache
+ * {% raw %}`myapp-{{ epoch }}`{% endraw %} - every build will generate separate cache
+
+While choosing suitable templates for your cache `key`, keep in mind that cache saving is not a free operation, because it will take some time to upload the cache to our storage. So it make sense to have a `key` that generates a new cache only if something actually changed and avoid generating a new one every build.
+
+## Tips and potential pitfalls with dependency caching
+
+### Partial cache restores are better than no cache restore
+
+In most cases, partial cache hits are preferable to zero cache hits. If you get a cache miss, you’ll have to reinstall all your dependencies, which can result in a significant performance hit.
```

Comment:
I think this depends in the circumstances.   Partial caches in python/php can be pretty bad, e.g. your code may end up depending on a dependency that's not inlined with your declared dependencies and the problem won't surface until you run a build without a cache.  Other languages build tools (e.g. java build tools) are smarter about it. 

If the dependencies change infrequently, I would suggest going for a zero cache hit first; and only supporting partial caching if the cache miss costs end up being significant over time.